lib/Runtime/Library/AtomicsObject.cpp

```diff
@@ -0,0 +1,270 @@
+        if (args.Info.Count <= length) \
+        { \
+            JavascriptError::ThrowRangeError(scriptContext, JSERR_Function_LessArguments, _u(methodName)); \
+        } \
+
+
+namespace Js
+{
+    Var AtomicsObject::ValidateSharedIntegerTypedArray(Var typedArray, ScriptContext *scriptContext, bool onlyInt32)
+    {
+        if (!TypedArrayBase::Is(typedArray))
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedTypedArrayObject);
+        }
+
+        if (onlyInt32)
+        {
+            if (!Int32Array::Is(typedArray))
+            {
+                JavascriptError::ThrowTypeError(scriptContext, JSERR_InvalidOperationOnTypedArray);
+            }
+        }
+        else
+        {
+            if (!(Int8Array::Is(typedArray) || Uint8Array::Is(typedArray) || Int16Array::Is(typedArray) ||
+                Uint16Array::Is(typedArray) || Int32Array::Is(typedArray) || Uint32Array::Is(typedArray)))
+            {
+                JavascriptError::ThrowTypeError(scriptContext, JSERR_InvalidOperationOnTypedArray);
+            }
+        }
+
+        TypedArrayBase *typedArrayBase = TypedArrayBase::FromVar(typedArray);
+        ArrayBufferBase* arrayBuffer = typedArrayBase->GetArrayBuffer();
+        if (arrayBuffer == nullptr || !ArrayBufferBase::Is(arrayBuffer) || !arrayBuffer->IsSharedArrayBuffer())
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedSharedArrayBufferObject);
+        }
+
+        return arrayBuffer;
+    }
+
+    uint32 AtomicsObject::ValidateAtomicAccess(Var typedArray, Var requestIndex, ScriptContext *scriptContext)
+    {
+        int32 accessIndex = -1;
+        if (TaggedInt::Is(requestIndex))
+        {
+            accessIndex = TaggedInt::ToInt32(requestIndex);
+        }
+        else
+        {
+            accessIndex = JavascriptConversion::ToInt32_Full(requestIndex, scriptContext);
+            double dblValue = JavascriptConversion::ToNumber(requestIndex, scriptContext);
+            if (dblValue != accessIndex)
+            {
+                JavascriptError::ThrowRangeError(scriptContext, JSERR_InvalidTypedArrayIndex);
+            }
+        }
+
+        Assert(TypedArrayBase::Is(typedArray));
+        
+        if (accessIndex < 0 || accessIndex >= (int32)TypedArrayBase::FromVar(typedArray)->GetLength())
+        {
+            JavascriptError::ThrowRangeError(scriptContext, JSERR_InvalidTypedArrayIndex);
+        }
+
+        return (uint32)accessIndex;
+    }
+
+    TypedArrayBase * AtomicsObject::ValidateAndGetTypedArray(Var typedArray, Var index, __out uint32 *accessIndex, ScriptContext *scriptContext, bool onlyInt32)
+    {
+        ValidateSharedIntegerTypedArray(typedArray, scriptContext, onlyInt32);
+        uint32 i = ValidateAtomicAccess(typedArray, index, scriptContext);
+        if (accessIndex != nullptr)
+        {
+            *accessIndex = i;
+        }
+        Assert(TypedArrayBase::Is(typedArray));
```

Comment:
> Assert(TypedArrayBase::Is(typedArray)); [](start = 7, length = 40)

TypedArrayBase::FromVar already have AssertMsg(Is(aValue), "must be a typed array"); no need for this, remove from all places